crates/ruff/src/rules/flake8_pyi/rules/redundant_literal_union.rs

```diff
@@ -0,0 +1,155 @@
+///
+/// A: Literal["A"] | str
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from typing import Literal
+///
+/// A: Literal["A"]
+/// ```
+#[violation]
+pub struct RedundantLiteralUnion {
+    literal: String,
+    builtin_type: ExprType,
+}
+
+impl Violation for RedundantLiteralUnion {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RedundantLiteralUnion {
+            literal,
+            builtin_type,
+        } = self;
+        format!("`Literal[{literal}]` is redundant in a union with `{builtin_type}`",)
+    }
+}
+
+/// PYI051
+pub(crate) fn redundant_literal_union<'a>(checker: &mut Checker, union: &'a Expr) {
+    let mut literal_exprs = Vec::new();
+    let mut builtin_types_in_union = FxHashSet::default();
+
+    // Adds a member to `literal_exprs` for each value in a `Literal`, and any builtin types
+    // to `builtin_types_in_union`.
+    let mut func = |expr: &'a Expr, _| {
+        if let Expr::Subscript(ast::ExprSubscript { value, slice, .. }) = expr {
+            if checker.semantic().match_typing_expr(value, "Literal") {
+                if let Expr::Tuple(ast::ExprTuple { elts, .. }) = slice.as_ref() {
+                    literal_exprs.extend(elts.iter());
+                } else {
+                    literal_exprs.push(slice);
+                }
+            }
+            return;
+        }
+
+        let Some(builtin_type) = match_builtin_type(expr, checker.semantic()) else {
+            return;
+        };
+        builtin_types_in_union.insert(builtin_type);
+    };
+
+    traverse_union(&mut func, checker.semantic(), union, None);
+
+    for literal_expr in literal_exprs {
+        let Some(constant_type) = match_constant_type(literal_expr) else {
+            continue;
+        };
+
+        if builtin_types_in_union.contains(&constant_type) {
+            checker.diagnostics.push(Diagnostic::new(
+                RedundantLiteralUnion {
+                    literal: checker.locator().slice(literal_expr.range()).to_string(),
+                    builtin_type: constant_type,
+                },
+                literal_expr.range(),
+            ));
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)]
+enum ExprType {
+    Int,
+    Str,
+    Bool,
+    Float,
+    Bytes,
+    Complex,
+}
+
+impl fmt::Display for ExprType {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Int => fmt.write_str("int"),
+            Self::Str => fmt.write_str("str"),
+            Self::Bool => fmt.write_str("bool"),
+            Self::Float => fmt.write_str("float"),
+            Self::Bytes => fmt.write_str("bytes"),
+            Self::Complex => fmt.write_str("complex"),
+        }
+    }
+}
+
+/// Return the [`ExprType`] of an [`Expr]` if it is a builtin type (e.g. `int`, `bool`, `float`,
+/// `str`, `bytes`, or `complex`).
+fn match_builtin_type(expr: &Expr, model: &SemanticModel) -> Option<ExprType> {
```

Comment:
@LaBatata101 - I think the issue is that we needed to treat these two cases as separate. Otherwise, we were flagging cases like `Literal[1, 2]`, since we'd detect that we have `int` in the `Union` (by way of visiting the `1` and `2`).